src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,293 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.ConvertToSwitch;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ConvertToSwitch
+{
+    internal sealed class CSharpConvertToSwitchCodeRefactoringProvider : AbstractConvertToSwitchCodeRefactoringProvider
+    {
+        private abstract class Pattern
+        {
+            public sealed class ByValue : Pattern
+            {
+                public readonly ExpressionSyntax Expression;
+                public ByValue(ExpressionSyntax expression)
+                {
+                    Expression = expression;
+                }
+            }
+            public sealed class ByType : Pattern
+            {
+                public readonly TypeSyntax Type;
+                public readonly SyntaxToken Identifier;
+                public ByType(TypeSyntax type, SyntaxToken identifier)
+                {
+                    Type = type;
+                    Identifier = identifier;
+                }
+            }
+            public sealed class Discarded : Pattern
+            {
+                public readonly TypeSyntax Type;
+                public Discarded(TypeSyntax type)
+                {
+                    Type = type;
+                }
+            }
+            public sealed class Guarded : Pattern
```

Comment:
A small code example of what sort of cases these individual patterns represent would be nice.